### PR TITLE
[v2][storage] Implement read path for v2 storage interface

### PR DESCRIPTION
**Pull request by mahadzaryab1**
## Which problem is this PR solving?
- Towards #5079

## Description of the changes
- Implemented the read path for the v2 storage interface. This path currently just wraps a v1 span reader and exposes a static method to access the v1 reader. 
- Change the jaeger query extension to initialize a v2 storage factory and obtain the v1 span reader from it. 
- This will unblock the development of more efficient v2 storage implementations, like ClickHouse. 

## How was this change tested?
- Added unit tests for new functionality

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

### CODE REVIEW

File: cmd/jaeger/internal/extension/jaegerquery/server.go [hunk 1]
```diff
@@ -55,18 +56,25 @@ func (s *server) Start(ctx context.Context, host component.Host) error {
...
+	}
+
+	spanReader, err := factoryadapter.GetV1Reader(traceReader)
 	if err != nil {
 		return fmt.Errorf("cannot create span reader: %w", err)
 	}
 
 	spanReader = storageMetrics.NewReadMetricsDecorator(spanReader, queryMetricsFactory)
 
+	// TODO: CreateDependencyReader currently not implemented
```
Comment (mahadzaryab1) [inline cmd/jaeger/internal/extension/jaegerquery/server.go:77]: @yurishkuro `CreateDependencyReader` isn't currently path of the v2 storage factory interface - how should we proceed here?

File: storage_v2/spanstore/factory.go [hunk 2]
```diff
@@ -17,4 +18,7 @@ type Factory interface {
 
 	// CreateTraceWriter creates a spanstore.Writer.
 	CreateTraceWriter() (Writer, error)
+
+	// CreateDependencyReader creates a dependencystore.Reader.
+	CreateDependencyReader() (dependencystore.Reader, error)
```
Comment (mahadzaryab1) [inline storage_v2/spanstore/factory.go:23]: @yurishkuro not sure if this belongs here since we're in `package spanstore` - let me know if you have any thoughts on how we should proceed here

File: cmd/jaeger/internal/extension/jaegerstorage/factoryadapter/reader.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+type TraceReader struct {
+	spanReader spanstore_v1.Reader
+}
+
+func GetV1Reader(reader spanstore.Reader) (spanstore_v1.Reader, error) {
+	tr, ok := reader.(*TraceReader)
+	if !ok {
+		return nil, ErrV1ReaderNotAvailable
+	}
+	return tr.spanReader, nil
```
Comment (yurishkuro) [inline cmd/jaeger/internal/extension/jaegerstorage/factoryadapter/reader.go:28]:
```suggestion
	if tr, ok := reader.(*TraceReader); ok {
		return tr.spanReader, nil
	}
	return nil, ErrV1ReaderNotAvailable
```

File: storage_v2/spanstore/factory.go (hunk 2 above)
Comment (yurishkuro) [reply storage_v2/spanstore/factory.go:23]: I think it needs to be a different interface

File: cmd/jaeger/internal/extension/jaegerquery/server.go [hunk 4]
```diff
@@ -55,12 +56,18 @@ func (s *server) Start(ctx context.Context, host component.Host) error {
...
 		return fmt.Errorf("cannot find primary storage %s: %w", s.config.Storage.TracesPrimary, err)
 	}
 
-	spanReader, err := f.CreateSpanReader()
+	traceReader, err := f.CreateTraceReader()
+	if err != nil {
+		return fmt.Errorf("cannot create trace reader: %w", err)
+	}
+
+	spanReader, err := factoryadapter.GetV1Reader(traceReader)
```
Comment (yurishkuro) [inline cmd/jaeger/internal/extension/jaegerquery/server.go:70]: the idea was to push this conversion down into query service (as deep as possible), otherwise we can never make it work with v2 API

Comment (mahadzaryab1) [reply cmd/jaeger/internal/extension/jaegerquery/server.go:70]: @yurishkuro Oh okay I see. So a couple of follow-ups 
- Do we want to take the `traceReader` as part of the constructor, hold it in the `QueryService` struct and then perform the conversion before calling the underlying method (in https://github.com/jaegertracing/jaeger/blob/main/cmd/query/app/querysvc/query_service.go#L63 for example)?
- Do we want to use the storage v2 factory in v1 context as well? The factory is initialized differently in v1 (example: https://github.com/jaegertracing/jaeger/blob/main/cmd/query/main.go#L42).

File: storage_v2/spanstore/factory.go (hunk 2 above)
Comment (mahadzaryab1) [reply storage_v2/spanstore/factory.go:23]: @yurishkuro Do we want to add an interface like https://github.com/jaegertracing/jaeger/blob/main/storage/dependencystore/interface.go#L1-L22 in `storage_v2` to operate on OTLP data and use the same trick to extract the v1 reader?

Comment (yurishkuro) [reply storage_v2/spanstore/factory.go]: no, I think we should keep that same interface for now. Dependencies are not part of OTLP, so we always need our own model, which we have in v1 interface, so I don't see a strong reason to create v2 of it.

Comment (mahadzaryab1) [reply storage_v2/spanstore/factory.go]: sounds good! i changed the storage extension to initialize v1 and v2 factories

File: cmd/jaeger/internal/extension/jaegerquery/server.go (hunk 4 above)
Comment (yurishkuro) [reply cmd/jaeger/internal/extension/jaegerquery/server.go:75]: > Do we want to take the traceReader as part of the constructor, hold it in the QueryService struct and then perform the conversion before calling the underlying method 

yes

> Do we want to use the storage v2 factory in v1 context as well?

Not quite following, what do you mean by context?

File: cmd/jaeger/internal/extension/jaegerquery/server.go [hunk 5]
```diff
@@ -66,6 +67,7 @@ func (s *server) Start(ctx context.Context, host component.Host) error {
 	}
 
 	spanReader = storageMetrics.NewReadMetricsDecorator(spanReader, queryMetricsFactory)
+	traceReader := factoryadapter.NewTraceReader(spanReader)
```
Comment (mahadzaryab1) [inline cmd/jaeger/internal/extension/jaegerquery/server.go:70]: @yurishkuro should we be getting the trace reader via `GetStorageFactoryV2`? I wasn't sure if that was sufficient because we decorate the spanreader above

Comment (yurishkuro) [reply cmd/jaeger/internal/extension/jaegerquery/server.go:70]: yes, now it should be calling GetStorageFactoryV2. And yes about the decorator. Also, we may want to use otel.MeterProvider in the decorator instead of metrics.Factory (we need to gradually start deprecating metrics.Factory - let's book a ticket for that)

File: storage_v2/factoryadapter/reader.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	"errors"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+
+	spanstore_v1 "github.com/jaegertracing/jaeger/storage/spanstore"
+	"github.com/jaegertracing/jaeger/storage_v2/spanstore"
+)
+
+var ErrV1ReaderNotAvailable = errors.New("v1 reader is not available")
```
Comment (yurishkuro) [inline storage_v2/factoryadapter/reader.go:17]: I am not yet seeing a need for this being a public variable
```suggestion
var errV1ReaderNotAvailable = errors.New("spanstore.Reader is not a wrapper around v1 reader")
```

File: cmd/jaeger/internal/extension/jaegerquery/server.go [hunk 7]
```diff
@@ -55,19 +54,24 @@ func (s *server) Start(ctx context.Context, host component.Host) error {
...
 
-	spanReader, err := f.CreateSpanReader()
+	v1Factory.CreateSpanReader()
+	traceReader, err := v2Factory.CreateTraceReader()
 	if err != nil {
 		return fmt.Errorf("cannot create span reader: %w", err)
 	}
 
-	spanReader = storageMetrics.NewReadMetricsDecorator(spanReader, queryMetricsFactory)
+	// TODO: decorate trace reader with metrics
```
Comment (mahadzaryab1) [inline cmd/jaeger/internal/extension/jaegerquery/server.go:72]: @yurishkuro How should we do this? The [decorator](https://github.com/jaegertracing/jaeger/blob/main/storage/spanstore/metrics/decorator.go) implements `spanstore.Reader` so if we create a new one on `spanstore_v2.Reader`, it won't work currently since we only ever call the underlying v1 methods. The other alternative could be is that we take we pass `metrics.Factory` (or `otel.MeterProvider`) into the factoryadapter and decorate the `spanReader` before wrapping it into a TraceReader. Let me know what you think.

File: cmd/jaeger/internal/extension/jaegerquery/server.go (hunk 5 above)
Comment (mahadzaryab1) [reply cmd/jaeger/internal/extension/jaegerquery/server.go]: @yurishkuro Changed to call `GetStorageFactoryV2`. For the deprecation, this is what `metrics.Factory` currently looks like:
```go
// Factory creates new metrics
type Factory interface {
	Counter(metric Options) Counter
	Timer(metric TimerOptions) Timer
	Gauge(metric Options) Gauge
	Histogram(metric HistogramOptions) Histogram

	// Namespace returns a nested metrics factory.
	Namespace(scope NSOptions) Factory
}
``` 
Looking at OTEL's [meter](https://github.com/open-telemetry/opentelemetry-go/blob/main/metric/meter.go), I see it has a counter, gauge, and histogram. I don't see a Timer - what would be the alternative for that? Also, how do we handle the nesting of namespaces?

File: cmd/jaeger/internal/extension/jaegerquery/server.go (hunk 7 above)
Comment (yurishkuro) [reply cmd/jaeger/internal/extension/jaegerquery/server.go:72]: I would like to defer this if possible - what if we use the existing v1 decorator before creating V2 adapter? I.e. I would wrap the span reader as soon as we get it from the v1 factory.

Comment (mahadzaryab1) [reply cmd/jaeger/internal/extension/jaegerquery/server.go:72]: @yurishkuro Do we still want to use `GetStorageFactoryV2`? If we do, then the tracereader is created from the spanreader there. The other approach would be to just initialize the trace reader here (which is what I was previously doing) using the factoryadapter rather than getting it from the v2 factory. What do you think?

File: cmd/jaeger/internal/extension/jaegerquery/server.go (hunk 5 above)
Comment (yurishkuro) [reply cmd/jaeger/internal/extension/jaegerquery/server.go]: Are you trying to implement metrics.Factory on top of OTEL MeterProvider? I think it would be great if we could do that, it would minimize our changes across the code base.

Timer doesn't have to be part of the factory API, it can be a util struct done on top of a Histogram - the only difference of timer from Histogram is that its Record() function is strongly typed to Duration.

Name spacing is generally handled by our implementations as a prefix on the metric name, so it would work just the same for OTEL Meter.

File: storage_v2/factoryadapter/factory.go [hunk 8]
```diff
@@ -35,8 +35,12 @@ func (f *Factory) Close(_ context.Context) error {
 }
 
 // CreateTraceReader implements spanstore.Factory.
-func (*Factory) CreateTraceReader() (spanstore.Reader, error) {
```
Comment (yurishkuro) [inline storage_v2/factoryadapter/factory.go:38]: can we move this change to another PR and merge it first?

Comment (yurishkuro) [reply storage_v2/factoryadapter/factory.go:38]: btw this would be a good use case for samling

Comment (mahadzaryab1) [reply storage_v2/factoryadapter/factory.go:38]: @yurishkuro https://github.com/jaegertracing/jaeger/pull/6221

File: cmd/jaeger/internal/extension/jaegerquery/server.go (hunk 5 above)
Comment (mahadzaryab1) [reply cmd/jaeger/internal/extension/jaegerquery/server.go]: @yurishkuro Would you able to clarify what you mean by `Are you trying to implement metrics.Factory on top of OTEL MeterProvider?` I was trying to see if we could remove `metrics.Factory` and just replace it with OTEL's Meter Provider. Is there a different approach you were thinking of?

Comment (yurishkuro) [reply cmd/jaeger/internal/extension/jaegerquery/server.go]: I was wrong, we already have `internal/metrics/otelmetrics/factory.go`

Comment (yurishkuro) [reply cmd/jaeger/internal/extension/jaegerquery/server.go]: I guess you were simply asking about v2 version of decorator that wraps Storage V2 API - the decorator can still work off our internal pkg/metrics API.

Comment (mahadzaryab1) [reply cmd/jaeger/internal/extension/jaegerquery/server.go]: sounds good - so do we need to book any work here?

Comment (yurishkuro) [reply cmd/jaeger/internal/extension/jaegerquery/server.go]: we just need not to drop the metrics for storage - this is related to https://github.com/jaegertracing/jaeger/issues/6219

File: cmd/jaeger/internal/extension/jaegerquery/server.go [hunk 9]
```diff
@@ -74,17 +74,21 @@ func (s *server) Start(ctx context.Context, host component.Host) error {
 		Namespace(metrics.NSOptions{Name: "jaeger"}).
 		Namespace(metrics.NSOptions{Name: "query"})
 
-	f, err := jaegerstorage.GetStorageFactory(s.config.Storage.TracesPrimary, host)
+	v1Factory, err := jaegerstorage.GetStorageFactory(s.config.Storage.TracesPrimary, host)
```
Comment (yurishkuro) [inline cmd/jaeger/internal/extension/jaegerquery/server.go:77]: add a comment why v1 is still needed (because of dependencies)

File: cmd/jaeger/internal/extension/jaegerquery/server.go [hunk 10]
```diff
@@ -74,17 +74,21 @@ func (s *server) Start(ctx context.Context, host component.Host) error {
...
 
-	spanReader, err := f.CreateSpanReader()
+	traceReader, err := f.CreateTraceReader()
 	if err != nil {
-		return fmt.Errorf("cannot create span reader: %w", err)
+		return fmt.Errorf("cannot create trace reader: %w", err)
 	}
 
-	depReader, err := f.CreateDependencyReader()
+	depReader, err := v1Factory.CreateDependencyReader()
```
Comment (yurishkuro) [inline cmd/jaeger/internal/extension/jaegerquery/server.go:91]: would it make sense to create `storage_v2/depstore/` before continuing, so that we don't have to keep this bifurcation?

**Review by yurishkuro — COMMENTED** [review]
a refactoring suggestion:
* storage_v/spanstore -> storage_v/tracestore

File: cmd/jaeger/internal/extension/jaegerquery/server.go (hunk 10 above)
Comment (mahadzaryab1) [reply cmd/jaeger/internal/extension/jaegerquery/server.go:91]: @yurishkuro Sure. What do we want to put in that package?

**Comment by mahadzaryab1** [comment]
> a refactoring suggestion:
> 
> * storage_v/spanstore -> storage_v/tracestore

Do we want to do this for both v1 and v2 or just v2?

**Comment by mahadzaryab1** [comment]
> a refactoring suggestion:
> 
> * storage_v/spanstore -> storage_v/tracestore

https://github.com/jaegertracing/jaeger/pull/6292

File: cmd/jaeger/internal/extension/jaegerquery/server.go [hunk 11]
```diff
@@ -72,17 +72,21 @@ func (s *server) Start(ctx context.Context, host component.Host) error {
...
 
-	f, err := jaegerstorage.GetStorageFactory(s.config.Storage.TracesPrimary, host)
+	v1Factory, err := jaegerstorage.GetStorageFactory(s.config.Storage.TracesPrimary, host)
 	if err != nil {
-		return fmt.Errorf("cannot find primary storage %s: %w", s.config.Storage.TracesPrimary, err)
+		return fmt.Errorf("cannot find primary storage from v1 factory %s: %w", s.config.Storage.TracesPrimary, err)
+	}
+	f, err := jaegerstorage.GetStorageFactoryV2(s.config.Storage.TracesPrimary, host)
+	if err != nil {
+		return fmt.Errorf("cannot find primary storage from v2 factory %s: %w", s.config.Storage.TracesPrimary, err)
```
Comment (yurishkuro) [inline cmd/jaeger/internal/extension/jaegerquery/server.go:81]:
```suggestion
		return fmt.Errorf("cannot find v2 factory for primary storage %s: %w", s.config.Storage.TracesPrimary, err)
```

File: cmd/query/app/querysvc/query_service.go [hunk 12]
```diff
@@ -40,15 +42,15 @@ type StorageCapabilities struct {
 
 // QueryService contains span utils required by the query-service.
 type QueryService struct {
-	spanReader       spanstore.Reader
+	traceReader      tracestore.Reader
```
Comment (yurishkuro) [inline cmd/query/app/querysvc/query_service.go:45]: what about archive reader/writer, any reason not to upgrade them too? (or PR is too large?)

File: cmd/query/app/querysvc/query_service.go [hunk 13]
```diff
@@ -61,7 +63,11 @@ func NewQueryService(spanReader spanstore.Reader, dependencyReader dependencysto
 
 // GetTrace is the queryService implementation of spanstore.Reader.GetTrace
 func (qs QueryService) GetTrace(ctx context.Context, traceID model.TraceID) (*model.Trace, error) {
-	trace, err := qs.spanReader.GetTrace(ctx, traceID)
+	spanReader, err := factoryadapter.GetV1Reader(qs.traceReader)
```
Comment (yurishkuro) [inline cmd/query/app/querysvc/query_service.go:66]: arguably, you could just call the tracer reader directly and transform the results into model - almost the same amount of code changes, but be done with it once and for all, no need to come back and revisit.

Comment (mahadzaryab1) [reply cmd/query/app/querysvc/query_service.go:66]: @yurishkuro what would this look like? we haven't implemented the v2 reader interface for the factoryadapter at the moment so how would we do this?

File: cmd/query/app/querysvc/query_service.go (hunk 12 above)
Comment (mahadzaryab1) [reply cmd/query/app/querysvc/query_service.go:45]: @yurishkuro I initially thought that we couldn't do this yet because I thought we were using the `ArchiveFactory` interface here. However, it seems like we just use the normal reader interface on the archive reader as well (see https://github.com/jaegertracing/jaeger/blob/main/cmd/query/app/querysvc/query_service.go#L69) so this can be done. I can do it in this PR or a separate one. Let me know if you have a preference.

File: cmd/query/app/querysvc/query_service.go (hunk 13 above)
Comment (yurishkuro) [reply cmd/query/app/querysvc/query_service.go:66]: ah, ok, I thought we did. We need to, but can be different PR.

File: cmd/query/app/querysvc/query_service.go (hunk 12 above)
Comment (yurishkuro) [reply cmd/query/app/querysvc/query_service.go:45]: Because of the other question (not having v2 reader fully implemented) it's ok to do these separately, no particular benefit afaict.

File: cmd/jaeger/internal/extension/jaegerquery/server.go [hunk 14]
```diff
@@ -72,17 +72,21 @@ func (s *server) Start(ctx context.Context, host component.Host) error {
 		Namespace(metrics.NSOptions{Name: "jaeger"}).
 		Namespace(metrics.NSOptions{Name: "query"})
 
-	f, err := jaegerstorage.GetStorageFactory(s.config.Storage.TracesPrimary, host)
+	v1Factory, err := jaegerstorage.GetStorageFactory(s.config.Storage.TracesPrimary, host)
```
Comment (yurishkuro) [inline cmd/jaeger/internal/extension/jaegerquery/server.go:75]:
```suggestion
	// TODO currently v1 is still needed because of dependency storage
	v1Factory, err := jaegerstorage.GetStorageFactory(s.config.Storage.TracesPrimary, host)
```

File: cmd/jaeger/internal/extension/jaegerquery/server.go [hunk 15]
```diff
@@ -72,17 +72,21 @@ func (s *server) Start(ctx context.Context, host component.Host) error {
...
 	if err != nil {
-		return fmt.Errorf("cannot find primary storage %s: %w", s.config.Storage.TracesPrimary, err)
+		return fmt.Errorf("cannot find primary storage from v1 factory %s: %w", s.config.Storage.TracesPrimary, err)
```
Comment (yurishkuro) [inline cmd/jaeger/internal/extension/jaegerquery/server.go:77]:
```suggestion
		return fmt.Errorf("cannot find v1 factory for primary storage %s: %w", s.config.Storage.TracesPrimary, err)
```

File: cmd/jaeger/internal/extension/jaegerquery/server.go (hunk 11 above)
Comment (yurishkuro) [inline cmd/jaeger/internal/extension/jaegerquery/server.go:81]:
```suggestion
		return fmt.Errorf("cannot find v2 factory for primary storage %s: %w", s.config.Storage.TracesPrimary, err)
```

File: cmd/jaeger/internal/extension/jaegerquery/server.go [hunk 16]
```diff
@@ -91,7 +95,7 @@ func (s *server) Start(ctx context.Context, host component.Host) error {
 	if err := s.addArchiveStorage(&opts, host); err != nil {
```
Comment (yurishkuro) [inline cmd/jaeger/internal/extension/jaegerquery/server.go:95]:
```suggestion
	// TODO archive storage still uses v1 factory
	if err := s.addArchiveStorage(&opts, host); err != nil {
```

**Review by yurishkuro — APPROVED** [review]
